crypto/s2n_aead_cipher_chacha20_poly1305.c

```diff
@@ -157,6 +148,106 @@ static int s2n_aead_chacha20_poly1305_destroy_key(struct s2n_session_key *key)
+    size_t out_len = 0;
+
+    GUARD_OSSL(EVP_AEAD_CTX_seal(key->evp_aead_ctx, out->data, &out_len, out->size, iv->data, iv->size, in->data, in_len, aad->data, aad->size), S2N_ERR_ENCRYPT);
+
+    S2N_ERROR_IF((in_len + S2N_TLS_CHACHA20_POLY1305_TAG_LEN) != out_len, S2N_ERR_ENCRYPT);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_decrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    gte_check(in->size, S2N_TLS_CHACHA20_POLY1305_TAG_LEN);
+    gte_check(out->size, in->size);
+    eq_check(iv->size, S2N_TLS_CHACHA20_POLY1305_IV_LEN);
+
+    size_t out_len = 0;
+
+    GUARD_OSSL(EVP_AEAD_CTX_open(key->evp_aead_ctx, out->data, &out_len, out->size, iv->data, iv->size, in->data, in->size, aad->data, aad->size), S2N_ERR_DECRYPT);
+
+    S2N_ERROR_IF((in->size - S2N_TLS_CHACHA20_POLY1305_TAG_LEN) != out_len, S2N_ERR_ENCRYPT);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_set_encryption_key(struct s2n_session_key *key, struct s2n_blob *in)
+{
+    eq_check(in->size, S2N_TLS_CHACHA20_POLY1305_KEY_LEN);
+
+    GUARD_OSSL(EVP_AEAD_CTX_init(key->evp_aead_ctx, EVP_aead_chacha20_poly1305(), in->data, in->size, S2N_TLS_CHACHA20_POLY1305_TAG_LEN, NULL), S2N_ERR_KEY_INIT);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_set_decryption_key(struct s2n_session_key *key, struct s2n_blob *in)
+{
+    eq_check(in->size, S2N_TLS_CHACHA20_POLY1305_KEY_LEN);
+
+    GUARD_OSSL(EVP_AEAD_CTX_init(key->evp_aead_ctx, EVP_aead_chacha20_poly1305(), in->data, in->size, S2N_TLS_CHACHA20_POLY1305_TAG_LEN, NULL), S2N_ERR_KEY_INIT);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_init(struct s2n_session_key *key)
+{
+    EVP_AEAD_CTX_zero(key->evp_aead_ctx);
+
+    return 0;
+}
+
+static int s2n_aead_chacha20_poly1305_destroy_key(struct s2n_session_key *key)
+{
+    EVP_AEAD_CTX_cleanup(key->evp_aead_ctx);
+
+    return 0;
+}
+
+#else /* No implementation exists for chosen cryptographic provider */
+
+static int s2n_aead_chacha20_poly1305_encrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    S2N_ERROR(S2N_ERR_ENCRYPT);
+}
+
+static int s2n_aead_chacha20_poly1305_decrypt(struct s2n_session_key *key, struct s2n_blob *iv, struct s2n_blob *aad, struct s2n_blob *in, struct s2n_blob *out)
+{
+    S2N_ERROR(S2N_ERR_DECRYPT);
+}
+
+static int s2n_aead_chacha20_poly1305_set_encryption_key(struct s2n_session_key *key, struct s2n_blob *in)
+{
+    S2N_ERROR(S2N_ERR_KEY_INIT);
+}
+
+static int s2n_aead_chacha20_poly1305_set_decryption_key(struct s2n_session_key *key, struct s2n_blob *in)
+{
+    S2N_ERROR(S2N_ERR_KEY_INIT);
+}
+
+static int s2n_aead_chacha20_poly1305_init(struct s2n_session_key *key)
+{
+    S2N_ERROR(S2N_ERR_KEY_INIT);
+}
+
+static int s2n_aead_chacha20_poly1305_destroy_key(struct s2n_session_key *key)
+{
+    S2N_ERROR(S2N_ERR_KEY_DESTROY);
+}
+
+#endif
```

Comment:
Do we need this `#else` with empty implementation? Would there be any issues if we did:

```
#if BoringSSL || AWSLC
// BoringSSL impl
#else 
// OpenSSL impl
#endif
```

Reply:
Just saw that you split this out this empty impl from the previous OpenSSL implementation that had internal `#ifdef`'s for OpenSSL 1.1.x, so having 3 implementations (BoringSSL/AWS-LC, OpenSSL 1.1.x, Empty/OpenSSL 1.0.x) makes more sense now.